### ISSUE TITLE
Improve torch.autograd.set_detect_anomaly documentation

### ISSUE THREAD

**Issue by enchainingrealm** [pytorch/pytorch#26408]
## 📚 Documentation

Training a model with `torch.autograd.set_detect_anomaly(True)` causes a severe memory leak because every line of code that is executed is stored in memory as a string. As far I as know, this memory leak isn't documented anywhere. The documentation for `set_detect_anomaly` should be updated with a warning.

cc @ezyang @gchanan @zou3519

**Comment by izdeby** [comment]
hi, @enchainingrealm, do you happen to have a repro for the memory leak?

**Comment by enchainingrealm** [comment]
Here is a minimal example of the memory leak. In the code below, we use a "model" that's just a single convolutional layer. If you run the code with `set_detect_anomaly(True)`, 4 strings are leaked on every "batch" (while loop iteration.) If you run the code with `set_detect_anomaly(False)`, nothing is leaked.

```
import torch
from pympler import tracker
from torch import nn


def main():
    model = nn.Conv2d(64, 64, 3, padding=1)
    data = torch.rand((1, 64, 256, 256))

    torch.autograd.set_detect_anomaly(True)
    the_tracker = tracker.SummaryTracker()

    while True:
        the_tracker.print_diff()
        data = model(data)
        the_tracker.print_diff()


if __name__ == "__main__":
    main()
```

**Comment by soumith** [comment]
cc: @albanD the original author of anomaly detection

**Comment by albanD** [comment]
Hi,

Two things:
- Your code leaks memory because `data = model(data)` in an infinite loop will create an infinitely large computational graph. I'm not sure how pympler works but it does not seem to be tracking the pytorch Tensors...
- The data stored by the anomaly_mode are stored inside the computational graph. So if you "leak" the computational graph, you leak the anomaly mode metadata.

You can see this by either saving the model result in another variable like `data2`, or by using `.detach()` on the result of your forward and that will solve your issue.



So no memory is leaked for sure but there definitely a memory (and performance) overhead associated with using this mode. Do you think this should be specified in the doc?

**Comment by enchainingrealm** [comment]
Yes, the memory and performance overhead should be specified in the documentation. Just from reading the current documentation, I saw no reason to disable anomaly detection, and that's why I left it on while training.

Further investigation revealed that my code is still leaking memory (albeit less severely) even with `set_detect_anomaly` disabled, because I forgot to call `.detach()` on the losses that I'm saving to a data structure.

**Comment by enchainingrealm** [comment]
Strangely, Pympler does detect leaked PyTorch tensors in the example below:

```
import torch

from pympler import tracker


def main():
    the_tracker = tracker.SummaryTracker()
    the_tracker.print_diff()

    tensor = torch.zeros(1000)

    the_tracker.print_diff()


if __name__ == "__main__":
    main()
```
